elephant/spectral.py

```diff
@@ -242,6 +242,121 @@ def welch_psd(signal, n_segments=8, len_segment=None,
+        sampling frequency should be given through keyword argument `fs`.
+    fs : float, optional
+        Specifies the sampling frequency of the input time series
+        Default: 1.0
+    nw : float, optional
+        Time bandwidth product
+        Default: 4.0
+    num_tapers : int, optional
+        Number of tapers used in 1. to obtain estimate of PSD. By default
+        [2*nw] - 1 is chosen.
+        Default: None
+    peak_resolution : float, optional
+        Desired frequency resolution of the obtained PSD estimate. When given
+        as a `float`, it is taken as frequency in Hz.
+        If None, it will be determined from other parameters.
+        Default: None.
+
+    Returns
+    -------
+    freqs : np.ndarray
+        Frequencies associated with power estimate in `psd`
+    psd : np.ndarray
+        PSD estimate of the time series in `signal`
+    """
+
+    # When the input is AnalogSignal, the data is added after rolling the axis
+    # for time index to the last
+    data = np.asarray(signal)
+    if isinstance(signal, neo.AnalogSignal):
+        data = np.rollaxis(data, 0, len(data.shape))
+
+    # If the data is given as AnalogSignal, use its attribute to specify the
+    # sampling frequency
+    if hasattr(signal, 'sampling_rate'):
+        fs = signal.sampling_rate.rescale('Hz').magnitude
+
+    # If fs and frequency resolution is pq Quantity get magnitude
+    if isinstance(fs, pq.quantity.Quantity):
+        fs = fs.rescale('Hz').magnitude
+    if isinstance(peak_resolution, pq.quantity.Quantity):
+        peak_resolution = peak_resolution.rescale('Hz').magnitude
+
+    # Number of data points in time series
+    if data.ndim == 1:
+        length_signal = np.shape(data)[0]
+    else:
+        length_signal = np.shape(data)[1]
+
+    # Determine time-halfbandwidth product from given parameters
+    if peak_resolution is not None:
+        if peak_resolution <= 0:
+            raise ValueError("peak_resolution must be positive")
+        else:
+            nw = length_signal / fs * peak_resolution / 2
+
+    if num_tapers is None:
+        num_tapers = np.floor(2*nw).astype(int) - 1
+    else:
+        if not isinstance(num_tapers, int):
+            raise TypeError("num_tapers must be integer")
+        elif num_tapers <= 0:
+            raise ValueError("num_tapers must be positive")
+
+    print(f'Number of tapers: {num_tapers}')
+
+    # Generate frequencies
+    freqs = np.fft.rfftfreq(length_signal, d=1/fs)
+
+    # Get slepian functions
+    slepain_fcts = scipy.signal.windows.dpss(M=length_signal,
+                                             NW=nw,
+                                             Kmax=num_tapers,
+                                             sym='False')
+
+    # Calculate approximately independent spectrum estimates
+    if data.ndim == 1:
+        tapered_signal = data * slepain_fcts
+    else:
+        tapered_signal = data[:, np.newaxis] * slepain_fcts
```

Comment:
Needs a comment. Also `slepain`.

Reply:
Broadcasting addressed in https://github.com/NeuralEnsemble/elephant/pull/417/commits/6cd853f9c99c3127b7e8743875e2a9403c404516. `slepain` was refactored in https://github.com/NeuralEnsemble/elephant/pull/417/commits/760d318f8c070d12164c24e1f593888e1ef87ab4.